File: 1849_splittingAStringIntoDescendingConsecutiveValues/main.cpp
#include <string>

class Solution {
private:
    bool backtrack(int startIdx, unsigned long long prev, std::string& s) {
        if(startIdx == s.size())
            return true;
        std::string curString = "";
        unsigned long long curVal;
        for(int endIdx = startIdx; endIdx < s.size(); ++endIdx) {
            curString.push_back(s[endIdx]);
            curVal = std::stoull(curString);
            if(curVal + 1 == prev && backtrack(endIdx+1, curVal, s))
                return true;
        }
        return false;
    }
public:
    bool splitString(std::string s) {
        std::string curString = "";
        unsigned long long curVal;
        // at most s.substr(0, s.size()-1)
        for(int i = 0; i < s.size()-1; ++i) {
            curString.push_back(s[i]);
            curVal = std::stoull(curString);
            if(backtrack(i+1, curVal, s))
                return true;
        }
        return false;
    }
};

File: 1849_splittingAStringIntoDescendingConsecutiveValues/main.py
class Solution:
    def splitString(self, s: str) -> bool:
        partition = []
        def backtrack(startIdx):
            if startIdx == len(s) and len(partition) >= 2:
                return True
            
            for endIdx in range(startIdx, len(s)):
                curValue = int(s[startIdx:endIdx+1])
                if len(partition) == 0 or partition[-1] == curValue + 1:
                    partition.append(curValue)
                    if backtrack(endIdx+1):
                        return True
                    partition.pop()
            return False
        
        return backtrack(0)
    

class Solution1:
    def splitString(self, s: str) -> bool:
        def backtrack(startIdx, prev):
            if startIdx == len(s):
                return True
            
            for endIdx in range(startIdx, len(s)):
                val = int(s[startIdx: endIdx+1])
                if val + 1 == prev and backtrack(endIdx+1, val):
                    return True
            return False
        
        for i in range(len(s)-1):
            val = int(s[:i+1])
            if backtrack(i+1, val):
                return True
        return False

print(Solution().splitString('050043'))
